Automattic/woocommerce-payments
# Ensure SEPA can work without CC payment method

Fixes https://github.com/Automattic/woocommerce-payments/issues/1615

#### Changes proposed in this Pull Request
This PR attempts to decouple "woocommerce_payments" from credit card so that "WooCommerce Payments" and "WooCommerce Payments - Credit Card" can be enabled/disabled separately. As well, SEPA, Sofort, Giropay can be enabled on its own. For example,

| WooCommerce Payments | Credit Card | SEPA | Result|
|-|-|-|-|
|:white_check_mark: | :x: | :x: | No payment methods| 
|:white_check_mark: | :white_check_mark: | :x: | Credit card available| 
|:white_check_mark: | :x: | :white_check_mark: | SEPA available| 
|:white_check_mark: | :white_check_mark: | :white_check_mark: | Credit Card + SEPA available| 
|:x: | :white_check_mark:  / :x: | :white_check_mark:  / :x:  | No payment methods| 

This PR will make changes to the existing hierarchy as follow:
```
WC_Settings_API [wc core]
  - WC_Payment_Gateway [wc core]
  	- WC_Payment_Gateway_CC  [wc core, credit card]
  		- WC_Payment_Gateway_WCPay [Gateway class for WooCommerce Payments] 
  			- WC_Payment_Gateway_WCPay_Subscriptions_Compat
  			- CC_Payment_Gateway [blank]
  			- Giropay_Payment_Gateway
  			- Sepa_Payment_Gateway
  			- Sofort_Payment_Gateway
```
to 
```
WC_Settings_API [wc core]
  - WC_Payment_Gateway [wc core]
  	- WC_Payment_Gateway_CC  [wc core, credit card]
  		- WC_Payment_Gateway_WCPay [Wrapper Gateway class for WooCommerce Payments] 
  			- CC_Payment_Gateway [Credit card payment method]
  				- WC_Payment_Gateway_WCPay_Subscriptions_Compat
  			- Giropay_Payment_Gateway
  			- Sepa_Payment_Gateway
  			- Sofort_Payment_Gateway
```

This PR proposes the following changes:
- Refactored `WC_Payment_Gateway_WCPay` class and pull out Credit Card's function into `CC_Payment_Gateway`.
- Update `WC_Payment_Gateway_WCPay`'s `GATEWAY_ID` from `woocommerce_payments` to a new ID, `woocommerce_payments_wrapper`. This aims to make `WC_Payment_Gateway_WCPay` behaves almost like an abstract class.
- Update `WC_Payment_Gateway_WCPay_Subscriptions_Compat` to extend from `CC_Payment_Gateway` instead of `WC_Payment_Gateway_WCPay`.
- Replace all `WC_Payment_Gateway_WCPay::GATEWAY_ID` to `CC_Payment_Gateway::GATEWAY_ID`
- Make sure Credit Card is still enabled by default when none of the euro debit payment methods is enabled (backward compatible).
- Renamed `tests/unit/payment-methods/test-class-wc-payment-gateway-wcpay.php` to `tests/unit/payment-methods/test-class-cc-payment-gateway.php`. This continue to test Credit Card payments.

#### Testing instructions
Test backward compatibility. 
1. Pull this branch
2. Check out an item. Credit Card should show by default. Same behaviour as `develop`.
3. WooCommerce Payment Methods settings (wp-admin/admin.php?page=wc-settings&tab=checkout) should continue to work
4. WooCommerce Payment setting should behaves the same (wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments)

Test with SEPA feature flag enabled
1. With the option _wcpay_feature_sepa = 1 
2. Go to WooCommerce Payment Methods settings (wp-admin/admin.php?page=wc-settings&tab=checkout), you should see just "WooCommerce Payments" method
![image](https://user-images.githubusercontent.com/572862/117212348-75afad80-adb7-11eb-89df-26fba042c269.png)
3. Click "Manage", you should see 2 more Payment Method checkboxes at the bottom; Credit Card and SEPA. 
![image](https://user-images.githubusercontent.com/572862/117212645-d6d78100-adb7-11eb-9c40-0cf95ca7e12c.png)
4. Uncheck credit card's and check SEPA's.
5. Check out an item, and only SEPA should show as payment method.

-------------------
